web3-providers-http: Migrate from xhr2-cookies to cross-fetch

ayanamidev opened the pull request:
closes #4868

I have replaced the old `xhr2-cookies` library to `cross-fetch` since it was better, and the `npm i` for `xhr2-cookies` was broken for the latest node LTS version as well.

Working perfectly like before.

Commit:
Migrate from xhr2-cookies to cross-fetch

ayanamidev commented on packages/web3-providers-http/src/index.js:
```diff
@@ -19,29 +19,33 @@
 
+// Apply missing polyfill for IE
+require('es6-promise').polyfill();
+require('abortcontroller-polyfill/dist/polyfill-patch-fetch');
```
I have added the polyfills in case of compatibility, but I think CI makes some complain about it

ayanamidev commented on packages/web3-providers-http/src/index.js:
```diff
@@ -19,29 +19,33 @@
     options = options || {};
 
-    this.withCredentials = options.withCredentials || false;
+    this.withCredentials = options.withCredentials;
```
This is no longer boolean  https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#sending_a_request_with_credentials_included

ayanamidev commented on packages/web3-providers-http/types/index.d.ts:
```diff
@@ -30,7 +30,6 @@ export interface HttpHeader {
 }
 
 export interface HttpProviderAgent {
-    baseUrl?: string;
```
Have removed it since this isn't supported by fetch as well as this is not a standard for Http Agent

ayanamidev commented on test/httpprovider.js:
```diff
@@ -5,55 +5,78 @@ var https = require('https');
+  return new Promise(function(resolve, reject) {
+    resolve(response);
+  });
+};
```
Simple version of fetch-mock function added

ayanamidev commented on test/httpprovider.js:
```diff
@@ -67,14 +90,15 @@ describe('web3-providers-http', function () {
 
             provider.send({}, function (err, result) {
-                assert.equal(typeof result, 'object');
+                assert.equal(typeof result, 'undefined');
```
This should not be an object as we didn't had any url to fetch

nazarhussain commented on packages/web3-providers-http/src/index.js:
```diff
@@ -52,37 +56,67 @@ var HttpProvider = function HttpProvider(host, options) {
+    }
+
+    // https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#sending_a_request_with_credentials_included
+    if (['include', 'same-origin', 'omit'].indexOf(this.withCredentials) !== -1) {
+      options.credentials = this.withCredentials;
+    }
```
Why only these credentials? Why any value provided from the user?

ayanamidev replied:
@nazarhussain This was the standard credentials documented from Mozilla as I have wrote the link of the documentation to the comments, should we allow this to any type of string?

nazarhussain commented on packages/web3-providers-http/src/index.js:
```diff
@@ -93,32 +127,50 @@ HttpProvider.prototype._prepareRequest = function(){
+        }
+
+        this.connected = true;
+        callback(errors.InvalidConnection(this.host));
```
Seems invalid case, how an invalid connection can be considered connected?

nazarhussain commented on packages/web3-providers-http/types/index.d.ts:
```diff
@@ -46,8 +45,9 @@ export interface HttpProviderOptions {
-    withCredentials: boolean;
+    withCredentials?: string;
     timeout: number;
+    timeoutId?: NodeJS.Timeout | number;
```
Do we need to export it to types? It should be a implementation detail not exposed to users.

ayanamidev replied:
@nazarhussain Right

nazarhussain commented on test/httpprovider.js:
```diff
@@ -5,55 +5,78 @@ var https = require('https');
+            var agent = new https.Agent();
+            var options = {agent: {https: agent}};
+            var provider = new HttpProvider('http://localhost:8545', options);
+            var result = await provider._prepareRequest();
```
Should not `_prepareRequest` create only a function which unless invoked should not return a promise?

ayanamidev replied:
@nazarhussain You mean we should return the function like () => fetch(this.host, options) instead of the promise itself?

nazarhussain replied:
I believe the `_prepareRequest` should not be `async`. It should return the request object synchronously. And later on we should be using `request.send()` to actually send and await. 

If that seems to be a big change, then we can rename the function to `_prepareAndSendRequest` that will make the context right.

Commit:
Address comments for web3-providers-http

ghost commented:
@nazarhussain Resolved those three comments

nikoulai reviewed (commented):
Please check the failing tests (except for e2e_windows)

Commit:
web3-providers-http: Prevent global leakage of this.connected

Commit:
Return server side error without creating additional error object

ghost commented:
@jdevcs @avkos @luu-alex @spacesailor24 @nazarhussain @nikoulai I have fixed the testcase, would like to have a look at?

Commit:
Fixed type of http.Agent

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export

Commit:
Removed unused variable

Commit:
Remove unnecessary catch block

nazarhussain reviewed (changes requested):
@ayanamidev Thanks for your contributions. Left one small comment. Rest seems good to me. 

There are two other points that need to address. 

1. We don't allow community contributions to change the `package-lock.json` file. So please remove the change from that file in your PR. Some team member will merge the PR and include the changes in `package-lock.json`. 
2. Add an entry to `CHANGELOG.md` describing what you are changing in this PR.

ghost commented:
@nazarhussain @nikoulai @spacesailor24 Would like to ask your thought about removing `_prepareRequest` function for `web3-providers-http`. This function was `xhr2-cookies` specific and doesn't export it to the wild. So I think it makes sense to remove with the shift to `cross-fetch` since it will only return promise with the current code.

Commit:
Remove unnecessary internal _prepareRequest function from web3-providers-http

ghost commented:
@nazarhussain @nikoulai @spacesailor24 Addressed comments

nazarhussain reviewed (changes requested):
@ayanamidev LGTM. Please add an entry into `CHANGELOG.md` describing the change.

Commit:
Update CHANGELOG.md

ghost commented:
@nazarhussain Oh, sorry fixed.

nazarhussain commented:
@ayanamidev Please resolve the conflicts. Those probably will be in CHANGELOG.md. Then I would be able to merge the PR right away.

ghost commented:
@nazarhussain Done. Every conflicts resolved without changes from source code
